src/workerd/api/urlpattern.c++

```diff
@@ -4,289 +4,191 @@
-    .port = this->port.map([](kj::String& str) { return kj::str(str); }),
-    .pathname = this->pathname.map([](kj::String& str) { return kj::str(str); }),
-    .search = this->search.map([](kj::String& str) { return kj::str(str); }),
-    .hash = this->hash.map([](kj::String& str) { return kj::str(str); }),
-    .baseUrl = this->baseURL.map([](kj::String& str) { return kj::str(str); }),
-  };
-}
+#define V(uppercase, lowercase)                                                                    \
+  kj::StringPtr URLPattern::get##uppercase() const {                                               \
+    auto value = inner.get_##lowercase();                                                          \
+    return kj::StringPtr(value.data(), value.size());                                              \
+  }
+URL_PATTERN_COMPONENTS(V)
+#undef V
 
 jsg::Ref<URLPattern> URLPattern::constructor(jsg::Lock& js,
-    jsg::Optional<URLPatternInput> input,
-    jsg::Optional<kj::String> baseURL,
-    jsg::Optional<URLPatternOptions> patternOptions) {
-  auto options = patternOptions.orDefault({});
-  KJ_SWITCH_ONEOF(kj::mv(input).orDefault(URLPatternInit{})) {
-    KJ_CASE_ONEOF(str, kj::String) {
-      KJ_SWITCH_ONEOF(jsg::UrlPattern::tryCompile(str.asPtr(),
-                          jsg::UrlPattern::CompileOptions{
-                            .baseUrl = baseURL.map([](kj::String& str) { return str.asPtr(); }),
-                            .ignoreCase = options.ignoreCase.orDefault(false),
-                          })) {
-        KJ_CASE_ONEOF(err, kj::String) {
-          JSG_FAIL_REQUIRE(TypeError, kj::mv(err));
-        }
-        KJ_CASE_ONEOF(pattern, jsg::UrlPattern) {
-          return create(js, kj::mv(pattern));
-        }
+    jsg::Optional<URLPatternInput> maybeInput,
+    jsg::Optional<kj::OneOf<kj::String, URLPatternOptions>> maybeBase,
+    jsg::Optional<URLPatternOptions> maybeOptions) {
+  ada::url_pattern_input input = ada::url_pattern_init{};
+  std::string_view* base = nullptr;
+  std::optional<ada::url_pattern_options> options{};
+
+  KJ_IF_SOME(i, maybeInput) {
+    KJ_SWITCH_ONEOF(i) {
+      KJ_CASE_ONEOF(str, kj::String) {
+        input = std::string_view(str.begin(), str.size());
+      }
+      KJ_CASE_ONEOF(init, URLPatternInit) {
+        input = init.toAdaType();
       }
     }
-    KJ_CASE_ONEOF(init, URLPatternInit) {
-      KJ_SWITCH_ONEOF(jsg::UrlPattern::tryCompile(init,
-                          jsg::UrlPattern::CompileOptions{
-                            .ignoreCase = options.ignoreCase.orDefault(false),
-                          })) {
-        KJ_CASE_ONEOF(err, kj::String) {
-          JSG_FAIL_REQUIRE(TypeError, kj::mv(err));
-        }
-        KJ_CASE_ONEOF(pattern, jsg::UrlPattern) {
-          return create(js, kj::mv(pattern));
-        }
+  }
+
+  KJ_IF_SOME(b, maybeBase) {
+    KJ_SWITCH_ONEOF(b) {
+      KJ_CASE_ONEOF(str, kj::String) {
+        *base = std::string_view(str.begin(), str.size());
```

Comment:
I don't see how this is not a null pointer deref, why even use raw pointers in the first place?

Reply:
ada::parse_url_pattern takes const std::string_view* pointer because it is more efficient compared to std::optional<std::string_view>

Reply:
That's fine, but the code is still a null deref as is.